src/substrait/function_registry.py

```diff
@@ -0,0 +1,378 @@
+
+
+def to_integer_option(txt: str):
+    if txt.isnumeric():
+        return ParameterizedType.IntegerOption(literal=int(txt))
+    else:
+        return ParameterizedType.IntegerOption(
+            parameter=ParameterizedType.IntegerParameter(name=txt)
+        )
+
+
+# TODO try using antlr grammar here as well
+def to_parameterized_type(dtype: str):
+    if "?" in dtype:
+        dtype = dtype.replace("?", "")
+        nullability = Type.NULLABILITY_NULLABLE
+    else:
+        nullability = Type.NULLABILITY_REQUIRED
+
+    if dtype == "boolean":
+        return ParameterizedType(bool=Type.Boolean(nullability=nullability))
+    elif dtype == "i8":
+        return ParameterizedType(i8=Type.I8(nullability=nullability))
+    elif dtype == "i16":
+        return ParameterizedType(i16=Type.I16(nullability=nullability))
+    elif dtype == "i32":
+        return ParameterizedType(i32=Type.I32(nullability=nullability))
+    elif dtype == "i64":
+        return ParameterizedType(i64=Type.I64(nullability=nullability))
+    elif dtype == "fp32":
+        return ParameterizedType(fp32=Type.FP32(nullability=nullability))
+    elif dtype == "fp64":
+        return ParameterizedType(fp64=Type.FP64(nullability=nullability))
+    elif dtype == "timestamp":
+        return ParameterizedType(timestamp=Type.Timestamp(nullability=nullability))
+    elif dtype == "timestamp_tz":
+        return ParameterizedType(timestamp_tz=Type.TimestampTZ(nullability=nullability))
+    elif dtype == "date":
+        return ParameterizedType(date=Type.Date(nullability=nullability))
+    elif dtype == "time":
+        return ParameterizedType(time=Type.Time(nullability=nullability))
+    elif dtype == "interval_year":
+        return ParameterizedType(
+            interval_year=Type.IntervalYear(nullability=nullability)
+        )
+    elif dtype.startswith("decimal") or dtype.startswith("DECIMAL"):
+        (_, precision, scale, _) = re.split(r"\W+", dtype)
+
+        return ParameterizedType(
+            decimal=ParameterizedType.ParameterizedDecimal(
+                scale=to_integer_option(scale),
+                precision=to_integer_option(precision),
+                nullability=nullability,
+            )
+        )
+    elif dtype.startswith("varchar"):
+        (_, length, _) = re.split(r"\W+", dtype)
+
+        return ParameterizedType(
+            varchar=ParameterizedType.ParameterizedVarChar(
+                length=to_integer_option(length), nullability=nullability
+            )
+        )
+    elif dtype.startswith("precision_timestamp"):
+        (_, precision, _) = re.split(r"\W+", dtype)
+
+        return ParameterizedType(
+            precision_timestamp=ParameterizedType.ParameterizedPrecisionTimestamp(
+                precision=to_integer_option(precision), nullability=nullability
+            )
+        )
+    elif dtype.startswith("precision_timestamp_tz"):
+        (_, precision, _) = re.split(r"\W+", dtype)
```

Comment:
If the type is malformed what happens?

Reply:
Removed the function, with the current impl users will probably see an antlr error